### PR TITLE
Cloudwatch Logs: Ignore non-time grouping fields in expressions and alerts

### LINKED ISSUE #63388: CloudWatch Logs: queries returning multiple time series cannot be used in alert rules
**What happened**:
CloudWatch Logs returning multiple time series cannot be used in alert rules. I can see this problem was reported in another issue's comment in the past (https://github.com/grafana/grafana/issues/46429#issuecomment-1185375094).

**What you expected to happen**:
Each time series returned in a CloudWatch logs query should be evaluated individually in an alert rule. 

**How to reproduce it (as minimally and precisely as possible)**:
Select the CloudWatch Logs data source and write a query returning multiple time series, like the following (if more than one log group is selected):
```
stats count(*) by bin(1h), @log
```
Each individual time series is shown correctly in a time series panel:

If the query above is used in an alert rule instead, the error `Failed to evaluate queries and expressions: input data must be a wide series but got type long (input refid)` pops up:

The alert rule works if the query returns a single time series:
```
stats count(*) by bin(1h)
```

**Anything else we need to know?**:
I have not been able to apply panel expressions (like `Reduce`) to CloudWatch Logs queries either: I always get an error. Maybe the issue is related?
Also, in case it's useful: I've checked the Query Inspector for the query above. No `labels` are populated, and `@log` is a textual field. 
```
{
  "response": {
    "results": {
      "A": {
        "status": 200,
        "frames": [
          {
            "schema": {
              "name": "/loggroup1",
              "refId": "A",
              "meta": {...}
              "fields": [
                {
                  "name": "bin(1h)",
                  "type": "time",
                  "typeInfo": {
                    "frame": "time.Time",
                    "nullable": true
                  },
                  "labels": {}
                },
                {
                  "name": "@log",
                  "type": "string",
                  "typeInfo": {
                    "frame": "string",
                    "nullable": true
                  },
                  "labels": {}
                },
                {
                  "name": "count(*)",
                  "type": "number",
                  "typeInfo": {
                    "frame": "float64",
                    "nullable": true
                  },
                  "labels": {}
                }
              ]
            },
...
```

**Environment**:
- Grafana version: tested in 9.3.1, 9.3.2, 9.3.6
- Data source type & version: Cloudwatch 5.0.0
- OS Grafana is installed on: Ubuntu 20.04
- User OS & Browser: Ubuntu 20.04, Firefox 109
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by iwysiu**
Fixes a bug where CloudWatch Logs queries using `group by` can't be alerted on/used in expressions. The problem was that when we send a group by query to CloudWatch, it tells us the groups by adding the grouping fields to each row.  Alerts and expressions expect there to be only numeric information in the rows, so having the grouping strings there causes to error when processing the frames for alerts and expressions. This PR removes the non-time grouping fields from synchronous logs queries (aka expression and alert queries). We could probably remove them from all queries if we wanted, but I wanted to minimize the change in case users were using those fields.

**Which issue(s) does this PR fix?**:

Fixes #63388 

**Special notes for your reviewer:**
I'm adding this to the 10.0.0 milestone because its a bug, but it's also kind of a feature since it's never worked? I can change it if we think we should.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by kevinwcyu — APPROVED** [review]

File: pkg/tsdb/cloudwatch/log_query_test.go [hunk 1]
```diff
@@ -538,7 +538,126 @@ func TestGroupingResultsWithNumericField(t *testing.T) {
 		},
 	}
 
-	groupedResults, err := groupResults(fakeDataFrame, []string{"httpresponse"})
+	groupedResults, err := groupResults(fakeDataFrame, []string{"httpresponse"}, false)
+	require.NoError(t, err)
+	assert.ElementsMatch(t, expectedGroupedFrames, groupedResults)
+}
+
+func TestGroupingResultsWithRemoveNonNumericTrue(t *testing.T) {
```
Comment (fridgepoet) [inline pkg/tsdb/cloudwatch/log_query_test.go:546]: This test was a bit tough to digest. I have some suggestions, feel free to take em or leave em!
* I tried to remove anything more than the “minimum” in order for the test to be more concise — tried not to remove something you were testing. but I also feel like you can probably get rid of some fakelog-b and reduce this even more…
* simplified the time so there’s no noise to read about parsing the format
* inlined some variables — goal is for reader to consume as much as possible as a whole from top to bottom

```
func TestGroupingResultsWithRemoveNonNumericTrue(t *testing.T) {
	logField := data.NewField(“@log”, data.Labels{}, []*string{
		aws.String(“fakelog-a”),
		aws.String(“fakelog-b”),
		aws.String(“fakelog-a”),
		aws.String(“fakelog-b”),
	})

	countField := data.NewField(“count”, data.Labels{}, []*string{
		aws.String(“100"),
		aws.String(“34”),
		aws.String(“57"),
		aws.String(“200”),
	})

	timeA := time.Time{}
	timeB := time.Time{}.Add(1 * time.Minute)
	fakeDataFrame := &data.Frame{
		Name: “CloudWatchLogsResponse”,
		Fields: []*data.Field{
			data.NewField(“@timestamp”, data.Labels{}, []*time.Time{&timeA, &timeA, &timeB, &timeB}),
			logField,
			countField,
		},
	}

	expectedGroupedFrames := []*data.Frame{
		{
			Name: “fakelog-a”,
			Fields: []*data.Field{
				data.NewField(“@timestamp”, data.Labels{}, []*time.Time{
					&timeA, &timeB,
				}),
				data.NewField(“count”, data.Labels{}, []*string{
					aws.String(“100”),
					aws.String(“57"),
				}),
			},
		},
		{
			Name: “fakelog-b”,
			Fields: []*data.Field{
				data.NewField(“@timestamp”, data.Labels{}, []*time.Time{
					&timeA, &timeB,
				}),
				data.NewField(“count”, data.Labels{}, []*string{
					aws.String(“34"),
					aws.String(“200”),
				}),
			},
		},
	}

	groupedResults, err := groupResults(fakeDataFrame, []string{“@log”}, true)
	require.NoError(t, err)
	assert.ElementsMatch(t, expectedGroupedFrames, groupedResults)
}
```

**Review by fridgepoet — APPROVED** [review]
Nice! Tested this out and am not getting the errors on alerting any more.
Since we are removing the same fields in removeFieldsByIndex and copyRowWithoutValues, I get the feeling we can remove the fields earlier… might be wrong there.
For the PR title, may need to be more about what’s done here -- “Ignore non-numeric and non-time fields in executeLogAction”?
Other thoughts but not necessarily for this PR:
* I’m wondering if we should try to untangle the parts of removing non-numeric/non-time fields outside of the groupResults function.
For example the unit test is already difficult to debug and for this feature we wouldn’t need to care about testing the grouping logic.
* groupResults is a bit bloated but has clear steps to its work -- may be nice to extract distinct parts of groupResults into their own functions which could make things clearer (won’t need as many of these comments explaining what parts are doing).

**Review by iwysiu — COMMENTED** [review]
As an update, I realized that the numeric grouping values get converted to strings also, and thus will probably have the same issue as the non-numeric grouping fields, so I'm also removing those.
For the bullets:
- We can’t remove them before groupResults function because the fields are the only information we have for which group a row belongs to, it’s how cloudwatch logs returns the data. I tried to simplify the test though!
- The steps in groupResults are interlinked enough that I'm not going to cut them apart now, but I could see that as future work. It's not so big that I have strong feelings about it.

**Comment by fridgepoet** [comment]
> I realized that the numeric grouping values get converted to strings also, and thus will probably have the same issue as the non-numeric grouping fields, so I'm also removing those

Could you explain this?

Also in the PR description you say "expressions expect there to be only numeric information in the rows" so is it only time now?

> the fields are the only information we have for which group a row belongs to
Don't we collect the rows in a map with groups for the keys, though?

**Comment by iwysiu** [comment]
@fridgepoet 
>Didn't understand this bit, could you explain this?
Which part of a unit test covers this detail?

To make the grouping keys for the grouped frames, `groupResults` converts the numeric fields to string fields (that first for loop in `groupResults`. I changed the stream field in the test to cover this case
>Also in the PR description you say "expressions expect there to be only numeric information in the rows." Is that supposed to say only time information in the rows now?

The non-grouping information will still be there, so count or whatever they're running will still be in the results. The number grouping fields were being returned as strings anyway, so those were not longer numeric
>Don't we collect the rows in a map with groups for the keys, though?

Yup! We do that in `groupResults`. Concievably we could change the field removal to a step after the grouping, but it can't really happen any earlier than it already is.

**Review by fridgepoet — COMMENTED** [review]
Oops I forgot to send some comments, also I don't see any latest changes after your last comment

File: pkg/tsdb/cloudwatch/log_query_test.go [hunk 2]
```diff
@@ -538,7 +538,72 @@ func TestGroupingResultsWithNumericField(t *testing.T) {
...
+					aws.String("62"),
+				}),
+			},
+			RefID: "",
+		},
+	}
+
+	groupedResults, err := groupResults(fakeDataFrame, []string{"@log", "stream"}, true)
 	require.NoError(t, err)
 	assert.ElementsMatch(t, expectedGroupedFrames, groupedResults)
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/log_query_test.go:608]: Neat! Unfortunately the frames aren't returned in a consistent order, so without further refactoring it's simpler to continue using ElementsMatch (which ignores ordering) for now.

**Review by iwysiu — COMMENTED** [review]
Good catch on the names!

**Pull request merged**